### PR TITLE
[MRG+1] Customize violin plot demo, see #6723

### PR DESCRIPTION
**Pull request by pd3**

### CODE REVIEW

**Comment by story645** [comment]
Just a tip. but if you wrote #6814 & #6814 in your description, than it automatically links.

**Comment by QuLogic** [comment]
Yes, please ensure you have run `pycodestyle` (formerly `pep8`) over the example as Travis is currently failing due to those issues.

File: examples/statistics/customized_violin_demo.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Customizing violin plots
+#
+#
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+# functions to calculate percentiles and adjacent values
+def percentile(vals, p):
```
Comment (tacaswell) [inline examples/statistics/customized_violin_demo.py]: why not use `np.percentile`?

File: examples/statistics/customized_violin_demo.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    if k >= N:
+        return vals[N-1]
+    return vals[k-1] + d*(vals[k] - vals[k-1])
+
+
+def adjacent_values(vals):
+    q1 = percentile(vals, 0.25)
+    q3 = percentile(vals, 0.75)
+    uav = q3 + (q3-q1)*1.5
+    if uav > vals[-1]:
```
Comment (tacaswell) [inline examples/statistics/customized_violin_demo.py]: can use `np.clip` here

File: examples/statistics/customized_violin_demo.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        lav = q1
+    return [lav, uav]
+
+# create test data
+dat = [np.random.normal(0, std, 100) for std in range(6, 10)]
+lab = ['a', 'b', 'c', 'd']    # labels
+med = []    # medians
+iqr = []    # inter-quantile ranges
+avs = []    # upper and lower adjacent values
+for arr in dat:
```
Comment (tacaswell) [inline examples/statistics/customized_violin_demo.py]: Much of this could be wrapped up in a function with a siganture like `def vln(ax, data, style)`

File: examples/statistics/customized_violin_demo.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    iqr.append([percentile(sarr, 0.25), percentile(sarr, 0.75)])
+    avs.append(adjacent_values(sarr))
+
+# plot the violins
+fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(7, 5))
+parts = ax.violinplot(dat, showmeans=False, showmedians=False,
+                      showextrema=False)
+
+# plot medians and averages
+for i in range(len(med)):
```
Comment (tacaswell) [inline examples/statistics/customized_violin_demo.py]: You can do the lines with a single call to `ax.vlines` and all of the markes with a single call to `ax.plot`.  On the other hand, if it is the idea to be able to make a single violin invisible  at a time this way might be better.

Can you collect the returned artists into a data structure of some sort (probably a dictionary per violin?)?

**Comment by tacaswell** [comment]
I am :+1: on this example over all, thank you for this!

I left some comments on how to make it better, but would not have any problem with this being merged as-is.

attn @story645

File: examples/statistics/customized_violin_demo.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+# customize colors
+for pc in parts['bodies']:
+    pc.set_facecolor('#D43F3A')
+    pc.set_edgecolor('black')
+    pc.set_alpha(1)
+
+ax.get_xaxis().set_tick_params(direction='out')
+ax.xaxis.set_ticks_position('bottom')
+ax.set_xticks([x+1 for x in range(len(lab))])
```
Comment (QuLogic) [inline examples/statistics/customized_violin_demo.py]: `np.arange(1, len(lab) + 1)`

**Comment by pd3** [comment]
@tacaswell I am not familiar with python enough though, please modify the example as necessary to adhere to your coding standards.

**Comment by story645** [comment]
I kind of would prefer it get simplified as I think some of the boilerplate is making it really hard to follow and sort of distracting from the core of the exame. If np.percentile and a call to vlines and the like work equally well,  I vote for that.

**Comment by pd3** [comment]
@story645 I do understand. Please modify the code as you deem fit, or otherwise ignore the pull request :-)

**Comment by story645** [comment]
Well it's your PR...I'm not trying to dictate anything here, just making a suggestion about what I think might be clearer-I could totally be wrong here and feel free to disagree. Honestly, I'm having an internal debate over my own code 'cause of all the `list(zip(*` constructions.

**Comment by NelleV** [comment]
The code is clean, and this is an nice addition to the gallery. I'd merged it as is, and do the modifications asked by @tacaswell and let @story645 improve the code in another PR (or I can put one of my students on it.)

**Comment by tacaswell** [comment]
Plan for this:

 - merge this as-is
 - create in issue pointing back here to apply some of the suggested changes

**Comment by tacaswell** [comment]
Backported to v2.x fc88474
